src/plugins/application_config/server/plugin.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Observable } from 'rxjs';
+import { first } from 'rxjs/operators';
+
+import {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+  IScopedClusterClient,
+  SharedGlobalConfig,
+} from '../../../core/server';
+
+import {
+  ApplicationConfigPluginSetup,
+  ApplicationConfigPluginStart,
+  ConfigurationClient,
+} from './types';
+import { defineRoutes } from './routes';
+import { OpenSearchConfigurationClient } from './opensearch_config_client';
+
+export class ApplicationConfigPlugin
+  implements Plugin<ApplicationConfigPluginSetup, ApplicationConfigPluginStart> {
+  private readonly logger: Logger;
+  private readonly config$: Observable<SharedGlobalConfig>;
+
+  private configurationClient: ConfigurationClient;
+  private configurationIndexName: string;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.logger = initializerContext.logger.get();
+    this.config$ = initializerContext.config.legacy.globalConfig$;
+    this.configurationIndexName = '';
+  }
+
+  private registerConfigurationClient(configurationClient: ConfigurationClient) {
+    this.logger.info('Register a configuration client.');
+    this.configurationClient = configurationClient;
+  }
+
+  private getConfigurationClient(configurationClient: IScopedClusterClient): ConfigurationClient {
+    if (this.configurationClient) {
+      return this.configurationClient;
+    }
+
+    const openSearchConfigurationClient = new OpenSearchConfigurationClient(
+      configurationClient,
+      this.configurationIndexName,
+      this.logger
+    );
+
+    return openSearchConfigurationClient;
+  }
+
+  public async setup(core: CoreSetup) {
+    const router = core.http.createRouter();
+
+    const config = await this.config$.pipe(first()).toPromise();
+
+    this.configurationIndexName = config.opensearchDashboards.configIndex;
+
+    // Register server side APIs
+    defineRoutes(router, this.getConfigurationClient.bind(this), this.logger);
+
+    return {
+      getConfigurationClient: this.getConfigurationClient.bind(this),
```

Comment:
Is this method needs to part of start? Have we tested this from plugin?

Reply:
1. 

Initially I thought about putting it in a `start`. However, for the Clickjacking use case, it needs this function to register a request handler during `setup`. See the draft PR for reference 

https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5641/files#diff-f72753e435f681570e82eb6935f6a95941e57c22b4bf6a42234c8f9370e80099R70

(Note that the content of the referred PR will updated greatly but the piece of logic to register request handler is still valid).

With this said, for other plugins who want to use the get function, they could still retrieve it during `setup` but could actually call it to perform real work during `start`. So we are not limiting the users.


2.

In terms of testing, I have tested it in the related PR. Also I have written an example storage plugin in current workspace with my current PR to just verify that a get function is retrieved from a plugin. The example code is checked in here https://github.com/tianleh/StorageProviderPOC/blob/main/server/plugin.ts#L34